pkg/daemon/ceph/client/upgrade.go

```diff
@@ -104,6 +157,197 @@ func EnableNautilusOSD(context *clusterd.Context) error {
+	if err != nil {
+		return fmt.Errorf("failed to check if %s was ok to stop. %+v", deployment, err)
+	}
+
+	return nil
+}
+
+// OkToContinue determines if it's ok to continue an upgrade
+func OkToContinue(context *clusterd.Context, namespace, deployment string) error {
+	daemonName, err := findDaemonName(deployment)
+	if err != nil {
+		logger.Warningf("%+v", err)
+		return nil
+	}
+
+	// the mon case is handled directly in the deployment where the mon checks for quorum
+	switch daemonName {
+	case "osd":
+		err := okToContinueOSDDaemon(context, namespace)
+		if err != nil {
+			return fmt.Errorf("failed to check if %s was ok to continue. %+v", deployment, err)
+		}
+	case "mds":
+		err := okToContinueMDSDaemon(context, namespace, deployment)
+		if err != nil {
+			return fmt.Errorf("failed to check if %s was ok to continue. %+v", deployment, err)
+		}
+	}
+
+	return nil
+}
+
+func okToStopDaemon(context *clusterd.Context, deployment, clusterName string) error {
+	daemonID := findDaemonID(deployment)
+	daemonName, err := findDaemonName(deployment)
+	if err != nil {
+		logger.Warningf("%+v", err)
+		return nil
+	}
+
+	if !stringInSlice(daemonName, daemonNoCheck) {
+		args := []string{daemonName, "ok-to-stop", daemonID}
+		command, args := FinalizeCephCommandArgs("ceph", args, context.ConfigDir, clusterName)
+
+		output, err := context.Executor.ExecuteCommandWithOutput(false, "", command, args...)
+		if err != nil {
+			return fmt.Errorf("deployment %s cannot be stopped. %+v", deployment, err)
+		}
+		logger.Debugf("deployment %s is ok to be updated. %s", deployment, output)
+	}
+
+	logger.Debugf("deployment %s is ok to be updated.", deployment)
+
+	return nil
+}
+
+// okToContinueOSDDaemon determines whether it's fine to go to the next osd during an upgrade
+// This basically makes sure all the PGs have settled
+func okToContinueOSDDaemon(context *clusterd.Context, namespace string) error {
+	// Reconciliating PGs should not take too long so let's wait up to 10 minutes
+	err := util.Retry(10, 60*time.Second, func() error {
+		return IsClusterClean(context, namespace)
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// okToContinueMDSDaemon determines whether it's fine to go to the next mds during an upgrade
+// mostly a placeholder function for the future but since we have standby mds this shouldn't be needed
+func okToContinueMDSDaemon(context *clusterd.Context, namespace, deployment string) error {
+	// wait for the MDS to be active again or in standby-replay
+	err := util.Retry(10, 15*time.Second, func() error {
+		return MdsActiveOrStandbyReplay(context, namespace, findFSName(deployment, namespace))
+	})
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
+
+func findFSName(deployment, namespace string) string {
+	mdsID := findDaemonID(deployment)
+	fsNameTrimSuffix := strings.TrimSuffix(deployment, "-"+mdsID)
+	return strings.TrimPrefix(fsNameTrimSuffix, namespace+"-mds-")
+}
+
+func findDaemonID(deployment string) string {
```

Comment:
I'm not sure this works for all daemons. The MDSes, for example are named `rook-ceph-mds-myfs-a` where `myfs-a` is the ID of the daemon. I do believe, however, that all daemons have their ID specified on their deployment by a label (except perhaps OSDs).

Also, not all daemons are controlled with deployments; the RGWs can optionally be daemonsets.

Reply:
bump